rustls/src/msgs/base.rs

```diff
@@ -148,50 +149,75 @@ impl fmt::Debug for PayloadU16 {
 }
 
 /// An arbitrary, unknown-content, u8-length-prefixed payload
+///
+/// `C` controls the minimum length accepted when decoding.
 #[derive(Clone, Eq, PartialEq)]
-pub struct PayloadU8(pub(crate) Vec<u8>);
+pub struct PayloadU8<C: Cardinality = MaybeEmpty>(pub(crate) Vec<u8>, PhantomData<C>);
 
-impl PayloadU8 {
+impl<C: Cardinality> PayloadU8<C> {
     pub(crate) fn encode_slice(slice: &[u8], bytes: &mut Vec<u8>) {
         (slice.len() as u8).encode(bytes);
         bytes.extend_from_slice(slice);
     }
 
     pub(crate) fn new(bytes: Vec<u8>) -> Self {
-        Self(bytes)
+        debug_assert!(bytes.len() >= C::MIN);
+        Self(bytes, PhantomData)
     }
+}
 
+impl PayloadU8<MaybeEmpty> {
     pub(crate) fn empty() -> Self {
-        Self(Vec::new())
+        Self(Vec::new(), PhantomData)
     }
 }
 
-impl Codec<'_> for PayloadU8 {
+impl<C: Cardinality> Codec<'_> for PayloadU8<C> {
     fn encode(&self, bytes: &mut Vec<u8>) {
+        debug_assert!(self.0.len() >= C::MIN);
         (self.0.len() as u8).encode(bytes);
         bytes.extend_from_slice(&self.0);
     }
 
     fn read(r: &mut Reader<'_>) -> Result<Self, InvalidMessage> {
         let len = u8::read(r)? as usize;
+        if len < C::MIN {
+            return Err(InvalidMessage::IllegalEmptyValue);
+        }
         let mut sub = r.sub(len)?;
         let body = sub.rest().to_vec();
-        Ok(Self(body))
+        Ok(Self(body, PhantomData))
     }
 }
 
-impl Zeroize for PayloadU8 {
+impl<C: Cardinality> Zeroize for PayloadU8<C> {
     fn zeroize(&mut self) {
         self.0.zeroize();
     }
 }
 
-impl fmt::Debug for PayloadU8 {
+impl<C: Cardinality> fmt::Debug for PayloadU8<C> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         hex(f, &self.0)
     }
 }
 
+pub trait Cardinality: Clone + Eq + PartialEq {
+    const MIN: usize;
+}
+
+#[derive(Clone, Eq, PartialEq)]
+pub struct MaybeEmpty;
+impl Cardinality for MaybeEmpty {
```

Comment:
nit: I think we'd usually want whitespace between the `struct` and `impl` here and on L216 no?